test/jdk/java/lang/Thread/crac/JoinOrSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,194 @@
+import jdk.test.lib.process.ProcessTools;
+
+import java.util.concurrent.CountDownLatch;
+
+
+public class JoinOrSleepOnCRPauseTest {
+
+    private static enum TestType {join, sleep};
+    private final TestType testType;
+
+    private final boolean nsAccuracy;
+
+    private final static long EPS_NS = Long.parseLong(
+        System.getProperty("cractests.time.eps", "100000000")); // default: 0.1s
+
+    private static final long T_MS = 2000;
+    private static final  int T_NS = 100;
+
+    private static final long CRPAUSE_MS = 4000;
+
+    private volatile long tDone = -1;
+
+    private final CountDownLatch checkpointLatch = new CountDownLatch(1);
+
+
+    private JoinOrSleepOnCRPauseTest(TestType testType, boolean nsAccuracy) {
+        this.testType   = testType;
+        this.nsAccuracy = nsAccuracy;
+    }
+
+    private void runTest() throws Exception {
+
+        String op = testType.name();
+
+        Thread mainThread = Thread.currentThread();
+
+        Runnable r = () -> {
+
+            try {
+
+                checkpointLatch.countDown();
+
+                switch (testType) {
+
+                    case join:
+                        if (nsAccuracy) {
+                            mainThread.join(T_MS, T_NS);
+                        } else {
+                            mainThread.join(T_MS);
+                        }
+                        break;
+
+                    case sleep:
+                        if (nsAccuracy) {
+                            Thread.sleep(T_MS, T_NS);
+                        } else {
+                            Thread.sleep(T_MS);
+                        }
+                        break;
+
+                    default:
+                        throw new IllegalArgumentException("unknown test type");
+                }
+
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+
+            tDone = System.nanoTime();
+        };
+
+        Thread t = new Thread(r);
+        t.start();
+
+        // this additional synchronization is probably redundant;
+        // adding it to ensure we get the expected TIMED_WAITING state
+        // from "our" join or sleep
+        checkpointLatch.await();
+
+        // it is expected that EPS_NS is enough to complete the join/sleep
+        // on restore => expecting that 5 * EPS_NS is enough to enter them
+        long dt = 5 * EPS_NS;
+        Thread.sleep(dt / 1_000_000);
+
+        if (t.getState() != Thread.State.TIMED_WAITING) {
+            throw new RuntimeException("was not able to enter " + op
+                + " in " + dt + " ns");
+        }
+
+        long tBeforeCheckpoint = System.nanoTime();
+
+        jdk.crac.Core.checkpointRestore();
+
+        long tAfterRestore = System.nanoTime();
+
+        t.join();
+
+        long pause = tAfterRestore - tBeforeCheckpoint;
+        if (pause < 1_000_000 * CRPAUSE_MS - EPS_NS) {
+            throw new RuntimeException(
```

Comment:
Please change to AssertionError here and below, when the expression should not fail

Reply:
done